notebooks/text_generator.livemd

```diff
@@ -0,0 +1,920 @@
+         [0.3723949193954468, 0.8734697103500366, 0.0638696625828743, -0.861497163772583, 2.1131083965301514, -1.5846097469329834, -0.4289300739765167, -0.5434455275535583, -1.112589955329895, 2.340670347213745, 1.0241754055023193, -1.0951896905899048, -1.0279741287231445, -0.8440263867378235, 0.06636204570531845, 0.8832864165306091, 0.47364047169685364, -0.7565221786499023, -0.0770493745803833, 1.1798070669174194, 7.5505499839782715, 0.11765921860933304, 1.1488550901412964, 1.1674405336380005, 0.14643093943595886, 2.751936435699463, -0.5080890655517578, 0.9232403039932251, 0.5222362279891968, 0.03229814022779465, -3.440410852432251, -0.44108346104621887, -0.17540469765663147, 1.124994158744812, 0.7630247473716736, 0.700600266456604, 2.5049147605895996, 0.7314786314964294, 0.6368493437767029, 1.2985477447509766, 0.4926140606403351, 0.7478111982345581, ...]
+       ]
+     >,
+     #Nx.Tensor<
+       f32[1][256]
+       EXLA.Backend<cuda:0, 0.1395165607.1903820818.68019>
+       [
+         [0.5909591317176819, 0.6029733419418335, 0.0828135758638382, 3.1577301025390625, -0.8707210421562195, 2.733414888381958, -0.3218168616294861, 0.2199571430683136, 1.3253854513168335, -0.7440951466560364, 0.7895384430885315, 0.013051019981503487, 1.1823228597640991, -0.36678698658943176, 0.06061827763915062, 1.4384580850601196, -0.1562608927488327, 0.4818105399608612, 0.15937495231628418, 0.14385631680488586, -5.125089645385742, 0.3450906276702881, 0.14982469379901886, 1.3046480417251587, 0.47652754187583923, -1.6058517694473267, -0.002407851628959179, 1.9648807048797607, 1.0021032094955444, 1.2189878225326538, -2.313563108444214, 0.44580522179603577, -0.1514829844236374, 1.8481801748275757, 1.4640742540359497, 0.2860425114631653, -1.201651930809021, -0.06269963085651398, -0.0290935467928648, 0.8209244608879089, -1.771423101425171, ...]
+       ]
+     >}
+  }
+}
+```
+
+## Generating Text
+
+After last step we will have our trained network parameters so we can start generating text with it. We simply need to give the initial sequence as the input to the network and select the most confident output. `Axon.predict/3` will give us the output layer and then using `Nx.argmax/1` we get the most confident neuron index then simply convert that index back to it's ASCII presentation.
+
+```elixir
+require Axon
+
+generate_fn = fn model, params, init_seq ->
+  # The initial sequence that we want the network to complete for us.
+  init_seq =
+    init_seq
+    |> String.trim()
+    |> String.downcase()
+    |> String.to_charlist()
+    |> Enum.map(&Map.fetch!(char_to_int, &1))
+
+  Enum.reduce(1..100, init_seq, fn _, seq ->
+    init_seq =
+      seq
+      |> Enum.take(-sequence_length)
+      |> Nx.tensor()
+      |> Nx.divide(characters_count)
+      |> Nx.reshape({1, sequence_length, 1})
+
+    char =
+      Axon.predict(model, params, init_seq)
+      |> Nx.argmax()
+      |> Nx.to_number()
+
+    seq ++ [char]
+  end)
+  |> Enum.map(&Enum.at(characters, &1))
```

Comment:
We can create `int_to_char` and use that for more efficient mapping.

Reply:
I created an `idx_to_char` mapping for explicitness. But AFAIK it wont help with efficiency as we only have `29` members and I believe BEAM treats maps smaller than `32` elements as lists. But it will make a difference if we expand the codepoints.

Reply:
> I believe BEAM treats maps smaller than 32 elements as lists

Correct, but the keys are sorted, so lookup via binary search is O(log N) :)